File: typescript_backup/src/webdriver-agent.ts
import { ActionableError, SwipeDirection, ScreenSize, ScreenElement, Orientation } from "./robot";

export interface SourceTreeElementRect {
	x: number;
	y: number;
	width: number;
	height: number;
}

export interface SourceTreeElement {
	type: string;
	label?: string;
	name?: string;
	value?: string;
	rawIdentifier?: string;
	rect: SourceTreeElementRect;
	isVisible?: string; // "0" or "1"
	children?: Array<SourceTreeElement>;
}

export interface SourceTree {
	value: SourceTreeElement;
}

export class WebDriverAgent {

	constructor(private readonly host: string, private readonly port: number) {
	}

	public async isRunning(): Promise<boolean> {
		const url = `http://${this.host}:${this.port}/status`;
		try {
			const response = await fetch(url);
			return response.status === 200;
		} catch (error) {
			console.error(`Failed to connect to WebDriverAgent: ${error}`);
			return false;
		}
	}

	public async createSession(): Promise<string> {
		const url = `http://${this.host}:${this.port}/session`;
		const response = await fetch(url, {
			method: "POST",
			headers: {
				"Content-Type": "application/json",
			},
			body: JSON.stringify({ capabilities: { alwaysMatch: { platformName: "iOS" } } }),
		});

		const json = await response.json();
		return json.value.sessionId;
	}

	public async deleteSession(sessionId: string) {
		const url = `http://${this.host}:${this.port}/session/${sessionId}`;
		const response = await fetch(url, { method: "DELETE" });
		return response.json();
	}

	public async withinSession(fn: (url: string) => Promise<any>) {
		const sessionId = await this.createSession();
		const url = `http://${this.host}:${this.port}/session/${sessionId}`;
		const result = await fn(url);
		await this.deleteSession(sessionId);
		return result;
	}

	public async getScreenSize(): Promise<ScreenSize> {
		return this.withinSession(async sessionUrl => {
			const url = `${sessionUrl}/wda/screen`;
			const response = await fetch(url);
			const json = await response.json();
			return {
				width: json.value.screenSize.width,
				height: json.value.screenSize.height,
				scale: json.value.scale || 1,
			};
		});
	}

	public async sendKeys(keys: string) {
		await this.withinSession(async sessionUrl => {
			const url = `${sessionUrl}/wda/keys`;
			await fetch(url, {
				method: "POST",
				headers: {
					"Content-Type": "application/json",
				},
				body: JSON.stringify({ value: [keys] }),
			});
		});
	}

	public async pressButton(button: string) {
		const _map = {
			"HOME": "home",
			"VOLUME_UP": "volumeup",
			"VOLUME_DOWN": "volumedown",
		};

		if (button === "ENTER") {
			await this.sendKeys("\n");
			return;
		}

		// Type assertion to check if button is a key of _map
		if (!(button in _map)) {
			throw new ActionableError(`Button "${button}" is not supported`);
		}

		await this.withinSession(async sessionUrl => {
			const url = `${sessionUrl}/wda/pressButton`;
			const response = await fetch(url, {
				method: "POST",
				headers: {
					"Content-Type": "application/json",
				},
				body: JSON.stringify({
					name: button,
				}),
			});

			return response.json();
		});
	}

	public async tap(x: number, y: number) {
		await this.withinSession(async sessionUrl => {
			const url = `${sessionUrl}/actions`;
			await fetch(url, {
				method: "POST",
				headers: {
					"Content-Type": "application/json",
				},
				body: JSON.stringify({
					actions: [
						{
							type: "pointer",
							id: "finger1",
							parameters: { pointerType: "touch" },
							actions: [
								{ type: "pointerMove", duration: 0, x, y },
								{ type: "pointerDown", button: 0 },
								{ type: "pause", duration: 100 },
								{ type: "pointerUp", button: 0 }
							]
						}
					]
				}),
			});
		});
	}

	private isVisible(rect: SourceTreeElementRect): boolean {
		return rect.x >= 0 && rect.y >= 0;
	}

	private filterSourceElements(source: SourceTreeElement): Array<ScreenElement> {
		const output: ScreenElement[] = [];

		const acceptedTypes = ["TextField", "Button", "Switch", "Icon", "SearchField", "StaticText", "Image"];

		if (acceptedTypes.includes(source.type)) {
			if (source.isVisible === "1" && this.isVisible(source.rect)) {
				if (source.label !== null || source.name !== null || source.rawIdentifier !== null) {
					output.push({
						type: source.type,
						label: source.label,
						name: source.name,
						value: source.value,
						identifier: source.rawIdentifier,
						rect: {
							x: source.rect.x,
							y: source.rect.y,
							width: source.rect.width,
							height: source.rect.height,
						},
					});
				}
			}
		}

		if (source.children) {
			for (const child of source.children) {
				output.push(...this.filterSourceElements(child));
			}
		}

		return output;
	}

	public async getPageSource(): Promise<SourceTree> {
		const url = `http://${this.host}:${this.port}/source/?format=json`;
		const response = await fetch(url);
		const json = await response.json();
		return json as SourceTree;
	}

	public async getElementsOnScreen(): Promise<ScreenElement[]> {
		const source = await this.getPageSource();
		return this.filterSourceElements(source.value);
	}

	public async openUrl(url: string): Promise<void> {
		await this.withinSession(async sessionUrl => {
			await fetch(`${sessionUrl}/url`, {
				method: "POST",
				body: JSON.stringify({ url }),
			});
		});
	}

        public async swipe(direction: SwipeDirection) {
                const screenSize = await this.getScreenSize();
                await this.withinSession(async sessionUrl => {

                        let x0: number, y0: number, x1: number, y1: number;
                        const centerX = screenSize.width >> 1;
                        const centerY = screenSize.height >> 1;

                        switch (direction) {
                                case "up":
                                        x0 = x1 = centerX;
                                        y0 = Math.floor(screenSize.height * 0.80);
                                        y1 = Math.floor(screenSize.height * 0.20);
                                        break;
                                case "down":
                                        x0 = x1 = centerX;
                                        y0 = Math.floor(screenSize.height * 0.20);
                                        y1 = Math.floor(screenSize.height * 0.80);
                                        break;
                                case "left":
                                        y0 = y1 = centerY;
                                        x0 = Math.floor(screenSize.width * 0.80);
                                        x1 = Math.floor(screenSize.width * 0.20);
                                        break;
                                case "right":
                                        y0 = y1 = centerY;
                                        x0 = Math.floor(screenSize.width * 0.20);
                                        x1 = Math.floor(screenSize.width * 0.80);
                                        break;
                                default:
                                        throw new ActionableError(`Swipe direction "${direction}" is not supported`);
                        }

			const url = `${sessionUrl}/actions`;
			await fetch(url, {
				method: "POST",
				headers: {
					"Content-Type": "application/json",
				},
				body: JSON.stringify({
					actions: [
						{
							type: "pointer",
							id: "finger1",
							parameters: { pointerType: "touch" },
							actions: [
								{ type: "pointerMove", duration: 0, x: x0, y: y0 },
								{ type: "pointerDown", button: 0 },
								{ type: "pointerMove", duration: 0, x: x1, y: y1 },
								{ type: "pause", duration: 1000 },
								{ type: "pointerUp", button: 0 }
							]
						}
					]
				}),
			});
                });
        }

        public async swipeBetweenPoints(startX: number, startY: number, endX: number, endY: number) {
                await this.withinSession(async sessionUrl => {
                        const url = `${sessionUrl}/actions`;
                        await fetch(url, {
                                method: "POST",
                                headers: { "Content-Type": "application/json" },
                                body: JSON.stringify({
                                        actions: [
                                                {
                                                        type: "pointer",
                                                        id: "finger1",
                                                        parameters: { pointerType: "touch" },
                                                        actions: [
                                                                { type: "pointerMove", duration: 0, x: startX, y: startY },
                                                                { type: "pointerDown", button: 0 },
                                                                { type: "pointerMove", duration: 0, x: endX, y: endY },
                                                                { type: "pause", duration: 1000 },
                                                                { type: "pointerUp", button: 0 }
                                                        ]
                                                }
                                        ]
                                })
                        });
                });
        }

	public async setOrientation(orientation: Orientation): Promise<void> {
		await this.withinSession(async sessionUrl => {
			const url = `${sessionUrl}/orientation`;
			await fetch(url, {
				method: "POST",
				headers: { "Content-Type": "application/json" },
				body: JSON.stringify({
					orientation: orientation.toUpperCase()
				})
			});
		});
	}

	public async getOrientation(): Promise<Orientation> {
		return this.withinSession(async sessionUrl => {
			const url = `${sessionUrl}/orientation`;
			const response = await fetch(url);
			const json = await response.json();
			return json.value.toLowerCase() as Orientation;
		});
	}
}

File: src/webdriver_agent.py
import json
from dataclasses import dataclass
from typing import Any, Awaitable, Callable, Dict, List, Optional

import aiohttp

from typing import Optional as OptionalType
from .robot import (
    ActionableError,
    Orientation,
    ScreenElement,
    ScreenElementRect,
    ScreenSize,
    SwipeDirection,
)


@dataclass
class SourceTreeElementRect:
    """소스 트리 요소의 위치 정보"""

    x: int
    y: int
    width: int
    height: int


@dataclass
class SourceTreeElement:
    """소스 트리 요소"""

    type: str
    rect: SourceTreeElementRect
    label: Optional[str] = None
    name: Optional[str] = None
    value: Optional[str] = None
    raw_identifier: Optional[str] = None
    is_visible: Optional[str] = None  # "0" or "1"
    children: Optional[List["SourceTreeElement"]] = None


@dataclass
class SourceTree:
    """소스 트리 루트"""

    value: SourceTreeElement


class WebDriverAgent:
    """iOS WebDriverAgent 클라이언트"""

    # 클래스 레벨 커넥터 (connection pool 재사용)
    _connector: Optional[aiohttp.TCPConnector] = None

    def __init__(self, host: str, port: int):
        self.host = host
        self.port = port
        self.base_url = f"http://{host}:{port}"

    @classmethod
    def _get_connector(cls) -> aiohttp.TCPConnector:
        """재사용 가능한 TCP 커넥터를 반환합니다."""
        if cls._connector is None or cls._connector.closed:
            cls._connector = aiohttp.TCPConnector(
                limit=10,  # 최대 동시 연결 수
                ttl_dns_cache=300,  # DNS 캐시 유지 시간
                keepalive_timeout=30,  # Keep-alive 타임아웃
            )
        return cls._connector

    def _create_session(self) -> aiohttp.ClientSession:
        """재사용 가능한 커넥터를 사용하는 세션을 생성합니다."""
        timeout = aiohttp.ClientTimeout(total=30)
        return aiohttp.ClientSession(
            connector=self._get_connector(),
            connector_owner=False,  # 세션 종료 시 커넥터 닫지 않음
            timeout=timeout,
        )

    async def is_running(self) -> bool:
        """WebDriverAgent가 실행 중인지 확인합니다."""
        url = f"{self.base_url}/status"
        try:
            async with self._create_session() as session:
                async with session.get(url) as response:
                    return response.status == 200
        except Exception as error:
            print(f"WebDriverAgent 연결 실패: {error}")
            return False

    async def create_session(self) -> str:
        """새 세션을 생성하고 세션 ID를 반환합니다."""
        url = f"{self.base_url}/session"

        async with self._create_session() as session:
            async with session.post(
                url, json={"capabilities": {"alwaysMatch": {"platformName": "iOS"}}}
            ) as response:
                data = await response.json()
                return data["value"]["sessionId"]

    async def delete_session(self, session_id: str) -> Dict[str, Any]:
        """세션을 삭제합니다."""
        url = f"{self.base_url}/session/{session_id}"

        async with self._create_session() as session:
            async with session.delete(url) as response:
                return await response.json()

    async def within_session(self, fn: Callable[[str], Awaitable[Any]]) -> Any:
        """세션 내에서 작업을 수행합니다."""
        session_id = await self.create_session()
        session_url = f"{self.base_url}/session/{session_id}"
        try:
            result = await fn(session_url)
            return result
        finally:
            await self.delete_session(session_id)

    async def get_screen_size(self) -> ScreenSize:
        """화면 크기를 가져옵니다."""

        async def _get_size(session_url: str) -> ScreenSize:
            url = f"{session_url}/wda/screen"
            async with self._create_session() as session:
                async with session.get(url) as response:
                    data = await response.json()
                    value = data["value"]
                    return ScreenSize(
                        width=value["screenSize"]["width"],
                        height=value["screenSize"]["height"],
                        scale=value.get("scale", 1),
                    )

        return await self.within_session(_get_size)

    async def send_keys(self, keys: str) -> None:
        """키 입력을 전송합니다."""

        async def _send(session_url: str) -> None:
            url = f"{session_url}/wda/keys"
            async with self._create_session() as session:
                await session.post(url, json={"value": [keys]})

        await self.within_session(_send)

    async def press_button(self, button: str) -> None:
        """버튼을 누릅니다."""
        button_map = {
            "HOME": "home",
            "VOLUME_UP": "volumeup",
            "VOLUME_DOWN": "volumedown",
        }

        if button == "ENTER":
            await self.send_keys("\n")
            return

        if button not in button_map:
            raise ActionableError(f'버튼 "{button}"은 지원되지 않습니다')

        async def _press(session_url: str) -> Dict[str, Any]:
            url = f"{session_url}/wda/pressButton"
            async with self._create_session() as session:
                async with session.post(url, json={"name": button_map[button]}) as response:
                    return await response.json()

        await self.within_session(_press)

    async def tap(self, x: int, y: int) -> None:
        """지정된 좌표를 탭합니다."""

        async def _tap(session_url: str) -> None:
            url = f"{session_url}/actions"
            actions = {
                "actions": [
                    {
                        "type": "pointer",
                        "id": "finger1",
                        "parameters": {"pointerType": "touch"},
                        "actions": [
                            {"type": "pointerMove", "duration": 0, "x": x, "y": y},
                            {"type": "pointerDown", "button": 0},
                            {"type": "pause", "duration": 100},
                            {"type": "pointerUp", "button": 0},
                        ],
                    }
                ]
            }

            async with self._create_session() as session:
                await session.post(url, json=actions)

        await self.within_session(_tap)

    async def double_tap(self, x: int, y: int) -> None:
        """지정된 좌표를 더블탭합니다."""

        async def _double_tap(session_url: str) -> None:
            url = f"{session_url}/actions"
            actions = {
                "actions": [
                    {
                        "type": "pointer",
                        "id": "finger1",
                        "parameters": {"pointerType": "touch"},
                        "actions": [
                            {"type": "pointerMove", "duration": 0, "x": x, "y": y},
                            {"type": "pointerDown", "button": 0},
                            {"type": "pause", "duration": 50},
                            {"type": "pointerUp", "button": 0},
                            {"type": "pause", "duration": 100},
                            {"type": "pointerDown", "button": 0},
                            {"type": "pause", "duration": 50},
                            {"type": "pointerUp", "button": 0},
                        ],
                    }
                ]
            }

            async with self._create_session() as session:
                await session.post(url, json=actions)

        await self.within_session(_double_tap)

    async def long_press(self, x: int, y: int, duration: OptionalType[int] = None) -> None:
        """지정된 좌표를 길게 누릅니다."""
        press_duration = duration if duration else 500  # 기본 500ms

        async def _long_press(session_url: str) -> None:
            url = f"{session_url}/actions"
            actions = {
                "actions": [
                    {
                        "type": "pointer",
                        "id": "finger1",
                        "parameters": {"pointerType": "touch"},
                        "actions": [
                            {"type": "pointerMove", "duration": 0, "x": x, "y": y},
                            {"type": "pointerDown", "button": 0},
                            {"type": "pause", "duration": press_duration},
                            {"type": "pointerUp", "button": 0},
                        ],
                    }
                ]
            }

            async with self._create_session() as session:
                await session.post(url, json=actions)

        await self.within_session(_long_press)

    def _is_visible(self, rect: SourceTreeElementRect) -> bool:
        """요소가 화면에 보이는지 확인합니다."""
        return rect.x >= 0 and rect.y >= 0

    def _filter_source_elements(self, source: SourceTreeElement) -> List[ScreenElement]:
        """소스 트리에서 화면 요소를 필터링합니다."""
        output: List[ScreenElement] = []

        accepted_types = [
            "TextField",
            "Button",
            "Switch",
            "Icon",
            "SearchField",
            "StaticText",
            "Image",
        ]

        if source.type in accepted_types:
            if source.is_visible == "1" and self._is_visible(source.rect):
                if source.label or source.name or source.raw_identifier:
                    output.append(
                        ScreenElement(
                            type=source.type,
                            label=source.label,
                            name=source.name,
                            value=source.value,
                            identifier=source.raw_identifier,
                            rect=ScreenElementRect(
                                x=source.rect.x,
                                y=source.rect.y,
                                width=source.rect.width,
                                height=source.rect.height,
                            ),
                        )
                    )

        if source.children:
            for child in source.children:
                output.extend(self._filter_source_elements(child))

        return output

    async def get_page_source(self) -> SourceTree:
        """페이지 소스를 가져옵니다."""
        url = f"{self.base_url}/source/?format=json"

        async with self._create_session() as session:
            async with session.get(url) as response:
                data = await response.json()
                return self._parse_source_tree(data)

    def _parse_source_tree(self, data: Dict[str, Any]) -> SourceTree:
        """JSON 데이터를 SourceTree로 파싱합니다."""

        def parse_element(elem_data: Dict[str, Any]) -> SourceTreeElement:
            rect_data = elem_data.get("rect", {})
            rect = SourceTreeElementRect(
                x=rect_data.get("x", 0),
                y=rect_data.get("y", 0),
                width=rect_data.get("width", 0),
                height=rect_data.get("height", 0),
            )

            children = None
            if "children" in elem_data:
                children = [parse_element(child) for child in elem_data["children"]]

            return SourceTreeElement(
                type=elem_data.get("type", ""),
                rect=rect,
                label=elem_data.get("label"),
                name=elem_data.get("name"),
                value=elem_data.get("value"),
                raw_identifier=elem_data.get("rawIdentifier"),
                is_visible=elem_data.get("isVisible"),
                children=children,
            )

        return SourceTree(value=parse_element(data["value"]))

    async def get_elements_on_screen(self) -> List[ScreenElement]:
        """화면의 모든 요소를 가져옵니다."""
        source = await self.get_page_source()
        return self._filter_source_elements(source.value)

    async def open_url(self, url: str) -> None:
        """URL을 엽니다."""

        async def _open(session_url: str) -> None:
            target_url = f"{session_url}/url"
            async with self._create_session() as session:
                await session.post(target_url, json={"url": url})

        await self.within_session(_open)

    async def swipe(self, direction: SwipeDirection) -> None:
        """스와이프합니다."""
        screen_size = await self.get_screen_size()

        async def _swipe(session_url: str) -> None:
            # 60% 스와이프 거리 사용 (mobile-mcp와 동일)
            vertical_distance = int(screen_size.height * 0.6)
            horizontal_distance = int(screen_size.width * 0.6)
            center_x = screen_size.width // 2
            center_y = screen_size.height // 2

            if direction == "up":
                x0 = x1 = center_x
                y0 = center_y + vertical_distance // 2
                y1 = center_y - vertical_distance // 2
            elif direction == "down":
                x0 = x1 = center_x
                y0 = center_y - vertical_distance // 2
                y1 = center_y + vertical_distance // 2
            elif direction == "left":
                y0 = y1 = center_y
                x0 = center_x + horizontal_distance // 2
                x1 = center_x - horizontal_distance // 2
            elif direction == "right":
                y0 = y1 = center_y
                x0 = center_x - horizontal_distance // 2
                x1 = center_x + horizontal_distance // 2
            else:
                raise ActionableError(f'스와이프 방향 "{direction}"은 지원되지 않습니다')

            url = f"{session_url}/actions"
            actions = {
                "actions": [
                    {
                        "type": "pointer",
                        "id": "finger1",
                        "parameters": {"pointerType": "touch"},
                        "actions": [
                            {"type": "pointerMove", "duration": 0, "x": x0, "y": y0},
                            {"type": "pointerDown", "button": 0},
                            {"type": "pointerMove", "duration": 1000, "x": x1, "y": y1},
                            {"type": "pointerUp", "button": 0},
                        ],
                    }
                ]
            }

            async with self._create_session() as session:
                resp = await session.post(url, json=actions)
                if not resp.ok:
                    error_text = await resp.text()
                    raise ActionableError(f"WebDriver actions request failed: {resp.status} {error_text}")
                # Clear actions to ensure they complete
                await session.delete(f"{session_url}/actions")

        await self.within_session(_swipe)

    async def swipe_between_points(
        self, start_x: int, start_y: int, end_x: int, end_y: int
    ) -> None:
        """지정된 좌표에서 다른 좌표까지 스와이프합니다."""

        async def _swipe(session_url: str) -> None:
            url = f"{session_url}/actions"
            actions = {
                "actions": [
                    {
                        "type": "pointer",
                        "id": "finger1",
                        "parameters": {"pointerType": "touch"},
                        "actions": [
                            {
                                "type": "pointerMove",
                                "duration": 0,
                                "x": start_x,
                                "y": start_y,
                            },
                            {"type": "pointerDown", "button": 0},
                            {
                                "type": "pointerMove",
                                "duration": 1000,
                                "x": end_x,
                                "y": end_y,
                            },
                            {"type": "pointerUp", "button": 0},
                        ],
                    }
                ]
            }

            async with self._create_session() as session:
                resp = await session.post(url, json=actions)
                if not resp.ok:
                    error_text = await resp.text()
                    raise ActionableError(f"WebDriver actions request failed: {resp.status} {error_text}")
                # Clear actions to ensure they complete
                await session.delete(f"{session_url}/actions")

        await self.within_session(_swipe)

    async def swipe_from_coordinate(
        self, x: int, y: int, direction: SwipeDirection, distance: OptionalType[int] = None
    ) -> None:
        """지정된 좌표에서 특정 방향으로 스와이프합니다."""
        # 기본 거리: 400 픽셀 (iOS 기본값)
        swipe_distance = distance if distance else 400

        x0 = x
        y0 = y
        x1 = x
        y1 = y

        if direction == "up":
            y1 = y - swipe_distance
        elif direction == "down":
            y1 = y + swipe_distance
        elif direction == "left":
            x1 = x - swipe_distance
        elif direction == "right":
            x1 = x + swipe_distance
        else:
            raise ActionableError(f'스와이프 방향 "{direction}"은 지원되지 않습니다')

        async def _swipe(session_url: str) -> None:
            url = f"{session_url}/actions"
            actions = {
                "actions": [
                    {
                        "type": "pointer",
                        "id": "finger1",
                        "parameters": {"pointerType": "touch"},
                        "actions": [
                            {"type": "pointerMove", "duration": 0, "x": x0, "y": y0},
                            {"type": "pointerDown", "button": 0},
                            {"type": "pointerMove", "duration": 1000, "x": x1, "y": y1},
                            {"type": "pointerUp", "button": 0},
                        ],
                    }
                ]
            }

            async with self._create_session() as session:
                resp = await session.post(url, json=actions)
                if not resp.ok:
                    error_text = await resp.text()
                    raise ActionableError(f"WebDriver actions request failed: {resp.status} {error_text}")
                # Clear actions to ensure they complete
                await session.delete(f"{session_url}/actions")

        await self.within_session(_swipe)

    async def set_orientation(self, orientation: Orientation) -> None:
        """화면 방향을 설정합니다."""

        async def _set(session_url: str) -> None:
            url = f"{session_url}/orientation"
            async with self._create_session() as session:
                await session.post(url, json={"orientation": orientation.upper()})

        await self.within_session(_set)

    async def get_orientation(self) -> Orientation:
        """현재 화면 방향을 가져옵니다."""

        async def _get(session_url: str) -> Orientation:
            url = f"{session_url}/orientation"
            async with self._create_session() as session:
                async with session.get(url) as response:
                    data = await response.json()
                    return data["value"].lower()

        return await self.within_session(_get)
